mobile_verifier/src/speedtests_average.rs

```diff
@@ -0,0 +1,463 @@
+                        latency_ms: st.latency as u32,
+                    })
+                    .collect(),
+                    validity: self.validity as i32,
+                    reward_multiplier: self.reward_multiplier.try_into().unwrap(),
+                },
+                &[("validity", self.validity.as_str_name())],
+            )
+            .await?;
+        Ok(())
+    }
+
+    pub fn reward_multiplier(&self) -> Decimal {
+        self.reward_multiplier
+    }
+
+    #[allow(dead_code)]
+    // function used by tests only
+    pub fn tier(&self) -> SpeedtestTier {
+        calculate_tier(
+            self.window_size,
+            self.upload_speed_avg_bps,
+            self.download_speed_avg_bps,
+            self.latency_avg_ms,
+        )
+    }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub enum SpeedtestTier {
+    Failed = 0,
+    Poor = 1,
+    Degraded = 2,
+    Acceptable = 3,
+}
+
+impl SpeedtestTier {
+    pub fn new(
+        window_size: usize,
+        upload_speed_avg_bps: u64,
+        download_speed_avg_bps: u64,
+        latency_avg_ms: u32,
+    ) -> SpeedtestTier {
+        calculate_tier(
+            window_size,
+            upload_speed_avg_bps,
+            download_speed_avg_bps,
+            latency_avg_ms,
+        )
+    }
+
+    fn into_multiplier(self) -> Decimal {
+        match self {
+            Self::Acceptable => dec!(1.0),
+            Self::Degraded => dec!(0.5),
+            Self::Poor => dec!(0.25),
+            Self::Failed => dec!(0.0),
+        }
+    }
+
+    fn from_download_speed(download_speed: u64) -> Self {
+        if download_speed >= mbps(100) {
+            Self::Acceptable
+        } else if download_speed >= mbps(50) {
+            Self::Degraded
+        } else if download_speed >= mbps(30) {
+            Self::Poor
+        } else {
+            Self::Failed
+        }
+    }
+
+    fn from_upload_speed(upload_speed: u64) -> Self {
+        if upload_speed >= mbps(10) {
+            Self::Acceptable
+        } else if upload_speed >= mbps(5) {
+            Self::Degraded
+        } else if upload_speed >= mbps(2) {
+            Self::Poor
+        } else {
+            Self::Failed
+        }
+    }
+
+    fn from_latency(latency: u32) -> Self {
+        if latency <= 50 {
+            Self::Acceptable
+        } else if latency <= 75 {
+            Self::Degraded
+        } else if latency <= 100 {
+            Self::Poor
+        } else {
+            Self::Failed
+        }
+    }
+}
+
+#[derive(Clone, Default)]
+pub struct SpeedtestAverages {
+    pub averages: HashMap<PublicKeyBinary, SpeedtestAverage>,
+}
+
+impl SpeedtestAverages {
+    pub fn get_average(&self, pub_key: &PublicKeyBinary) -> Option<SpeedtestAverage> {
+        self.averages.get(pub_key).cloned()
+    }
+
+    pub async fn aggregate_epoch_averages<'a>(
+        epoch_end: DateTime<Utc>,
+        exec: impl sqlx::PgExecutor<'a> + Copy + 'a,
```

Comment:
This essentially force the executor to be a `PgPool`, so you might as well just make it that to clean up the argument

Reply:
addressed here: a9b971c45caf62acb2c7b3c439ab824bad3ba01e
NOTE: heartbeats uses the same approach with the executor ( via HeartbeatReward::validated )